kernel/ide.c

```diff
@@ -148,21 +148,39 @@ int write_sector (uint64_t sector, const void *from) {
 
   outb(bar4 + ATA_BUSMASTER_REG_PRIMARY_COMMAND, ATA_BUSMASTER_CMD_START);
 
-  uint8_t status;
-  while (1) {
-    status = inb(bar4 + ATA_BUSMASTER_REG_PRIMARY_STATUS);
-    if (ATA_BUSMASTER_SR_SENT_IRQ & status) {
-      outb(bar4 + ATA_BUSMASTER_REG_PRIMARY_STATUS, ATA_BUSMASTER_SR_SENT_IRQ);
-      outb(bar4 + ATA_BUSMASTER_REG_PRIMARY_COMMAND, 0x00);
-      break;
-    }
+  operating_tcb = running_tcb;
+  is_read = 0;
+  schedule();
+}
+
+void ide_handler (void) {
+  uint8_t status = inb(bar4 + ATA_BUSMASTER_REG_PRIMARY_STATUS);
+  if (!(ATA_BUSMASTER_SR_SENT_IRQ & status)) {
+    panic("Bad IDE IRQ!");
   }
+  outb(bar4 + ATA_BUSMASTER_REG_PRIMARY_STATUS, ATA_BUSMASTER_SR_SENT_IRQ);
+  outb(bar4 + ATA_BUSMASTER_REG_PRIMARY_COMMAND, 0x00);
   if (ATA_BUSMASTER_SR_FAIL & status) {
-    outb(bar4 + ATA_BUSMASTER_REG_PRIMARY_STATUS, ATA_BUSMASTER_SR_FAIL);
-    goto err;
+    panic("Failed IDE IRQ!");
```

Comment:
This doesn't seem panic worthy when you can just return failure.

Reply:
Hmm, I guess I could write that error into the responsible thread's return value.  I'll do that.